{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 250)\n",
    "# Don't cut off long string\n",
    "# http://stackoverflow.com/questions/26277757/pandas-to-html-truncates-string-contents\n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "# for auto-reloading external modules\n",
    "# see http://stackoverflow.com/questions/1907993/autoreload-of-modules-in-ipython\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extracting mapping information from annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ndf1 = pd.read_pickle('../download/ensembl/Homo_sapiens.GRCh37.75.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 1 seqname                      2 source                       3 feature                     \n",
      " 4 start                        5 end                          6 score                       \n",
      " 7 strand                       8 frame                        9 ccds_id                     \n",
      "10 exon_id                     11 exon_number                 12 gene_biotype                \n",
      "13 gene_id                     14 gene_name                   15 gene_source                 \n",
      "16 protein_id                  17 tag:CCDS                    18 tag:cds_end_NF              \n",
      "19 tag:cds_start_NF            20 tag:mRNA_end_NF             21 tag:mRNA_start_NF           \n",
      "22 tag:seleno                  23 transcript_id               24 transcript_name             \n",
      "25 transcript_source           "
     ]
    }
   ],
   "source": [
    "_cols = ndf1.columns.tolist()\n",
    "for k, i in enumerate(_cols):\n",
    "    if k % 3 == 0 and k > 0:\n",
    "        print()\n",
    "    print('{0:2d} {1:28s}'.format(k + 1, i), end='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tag:CCDS</th>\n",
       "      <th>tag:cds_end_NF</th>\n",
       "      <th>tag:cds_start_NF</th>\n",
       "      <th>tag:mRNA_end_NF</th>\n",
       "      <th>tag:mRNA_start_NF</th>\n",
       "      <th>tag:seleno</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>havana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSE00002234944</td>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>havana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSE00003582793</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>havana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSE00002312635</td>\n",
       "      <td>3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>havana</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname                source     feature  start    end score strand frame ccds_id          exon_id exon_number gene_biotype          gene_id gene_name     gene_source protein_id  tag:CCDS  tag:cds_end_NF  tag:cds_start_NF  tag:mRNA_end_NF  tag:mRNA_start_NF  tag:seleno    transcript_id  \\\n",
       "0  1       pseudogene            gene        11869  14412  .     +      .     NaN     NaN              NaN         pseudogene   ENSG00000223972  DDX11L1   ensembl_havana  NaN       NaN       NaN             NaN               NaN              NaN                NaN          NaN               \n",
       "1  1       processed_transcript  transcript  11869  14409  .     +      .     NaN     NaN              NaN         pseudogene   ENSG00000223972  DDX11L1   ensembl_havana  NaN       NaN       NaN             NaN               NaN              NaN                NaN          ENST00000456328   \n",
       "2  1       processed_transcript  exon        11869  12227  .     +      .     NaN     ENSE00002234944  1           pseudogene   ENSG00000223972  DDX11L1   ensembl_havana  NaN       NaN       NaN             NaN               NaN              NaN                NaN          ENST00000456328   \n",
       "3  1       processed_transcript  exon        12613  12721  .     +      .     NaN     ENSE00003582793  2           pseudogene   ENSG00000223972  DDX11L1   ensembl_havana  NaN       NaN       NaN             NaN               NaN              NaN                NaN          ENST00000456328   \n",
       "4  1       processed_transcript  exon        13221  14409  .     +      .     NaN     ENSE00002312635  3           pseudogene   ENSG00000223972  DDX11L1   ensembl_havana  NaN       NaN       NaN             NaN               NaN              NaN                NaN          ENST00000456328   \n",
       "\n",
       "  transcript_name transcript_source  \n",
       "0  NaN             NaN               \n",
       "1  DDX11L1-002     havana            \n",
       "2  DDX11L1-002     havana            \n",
       "3  DDX11L1-002     havana            \n",
       "4  DDX11L1-002     havana            "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ndf2 = pd.read_pickle('../download/ensembl/Homo_sapiens.GRCh38.93.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 1 seqname                      2 source                       3 feature                     \n",
      " 4 start                        5 end                          6 score                       \n",
      " 7 strand                       8 frame                        9 ccds_id                     \n",
      "10 exon_id                     11 exon_number                 12 exon_version                \n",
      "13 gene_biotype                14 gene_id                     15 gene_name                   \n",
      "16 gene_source                 17 gene_version                18 protein_id                  \n",
      "19 protein_version             20 tag:CCDS                    21 tag:basic                   \n",
      "22 tag:cds_end_NF              23 tag:cds_start_NF            24 tag:mRNA_end_NF             \n",
      "25 tag:mRNA_start_NF           26 tag:seleno                  27 transcript_biotype          \n",
      "28 transcript_id               29 transcript_name             30 transcript_source           \n",
      "31 transcript_support_level    32 transcript_version          "
     ]
    }
   ],
   "source": [
    "_cols = ndf2.columns.tolist()\n",
    "for k, i in enumerate(_cols):\n",
    "    if k % 3 == 0 and k > 0:\n",
    "        print()\n",
    "    print('{0:2d} {1:28s}'.format(k + 1, i), end='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "s1 = set(ndf1.columns.values)\n",
    "s2 = set(ndf2.columns.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s1.difference(s2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon_version\n",
      "gene_version\n",
      "protein_version\n",
      "tag:basic\n",
      "transcript_biotype\n",
      "transcript_support_level\n",
      "transcript_version\n"
     ]
    }
   ],
   "source": [
    "for i in sorted(s2.difference(s1)):print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>tag:CCDS</th>\n",
       "      <th>tag:basic</th>\n",
       "      <th>tag:cds_end_NF</th>\n",
       "      <th>tag:cds_start_NF</th>\n",
       "      <th>tag:mRNA_end_NF</th>\n",
       "      <th>tag:mRNA_start_NF</th>\n",
       "      <th>tag:seleno</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>transcript_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>havana</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00002234944</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>havana</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00003582793</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>havana</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td>ENSE00002312635</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>havana</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>havana</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end score strand frame ccds_id          exon_id exon_number exon_version                        gene_biotype          gene_id gene_name gene_source gene_version protein_id protein_version tag:CCDS tag:basic tag:cds_end_NF tag:cds_start_NF tag:mRNA_end_NF  \\\n",
       "0  1       havana  gene        11869  14409  .     +      .                                                       transcribed_unprocessed_pseudogene  ENSG00000223972  DDX11L1   havana      5                                                                                                           \n",
       "1  1       havana  transcript  11869  14409  .     +      .                                                       transcribed_unprocessed_pseudogene  ENSG00000223972  DDX11L1   havana      5                                                1                                                          \n",
       "2  1       havana  exon        11869  12227  .     +      .             ENSE00002234944  1           1            transcribed_unprocessed_pseudogene  ENSG00000223972  DDX11L1   havana      5                                                1                                                          \n",
       "3  1       havana  exon        12613  12721  .     +      .             ENSE00003582793  2           1            transcribed_unprocessed_pseudogene  ENSG00000223972  DDX11L1   havana      5                                                1                                                          \n",
       "4  1       havana  exon        13221  14409  .     +      .             ENSE00002312635  3           1            transcribed_unprocessed_pseudogene  ENSG00000223972  DDX11L1   havana      5                                                1                                                          \n",
       "\n",
       "  tag:mRNA_start_NF tag:seleno    transcript_biotype    transcript_id transcript_name transcript_source transcript_support_level transcript_version  \n",
       "0                                                                                                                                                    \n",
       "1                               processed_transcript  ENST00000456328  DDX11L1-202     havana            1                        2                  \n",
       "2                               processed_transcript  ENST00000456328  DDX11L1-202     havana            1                        2                  \n",
       "3                               processed_transcript  ENST00000456328  DDX11L1-202     havana            1                        2                  \n",
       "4                               processed_transcript  ENST00000456328  DDX11L1-202     havana            1                        2                  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf2.head().fillna('')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
